### PR TITLE
Implement reportDoubleVote custom action

### PR DESCRIPTION
**Pull request by HoOngEe**

### CODE REVIEW

**Comment by majecty** [comment]
@HoOngEe What is the current status of this PR?
Do you need someone to review it? Or something is not done yet?

**Comment by HoOngEe** [comment]
@majecty I need someone to review it.

**Comment by majecty** [comment]
@HoOngEe Please fix the conflicts.
After fixing it, @foriequal0 and I will review this PR.

**Comment by HoOngEe** [comment]
@majecty I resolved the conflicts

File: core/src/consensus/stake/actions.rs [hunk 1]
```diff
@@ -102,12 +109,42 @@ impl Action {
...
+                let signer_idx1 = message1.signer_index();
+                let signer_idx2 = message2.signer_index();
+
+                let signers = signers.ok_or_else(|| {
+                    SyntaxError::InvalidCustomAction(String::from(
+                        "ReportDoubleVote Custom action is invalid for consensus engines without validators",
+                    ))
+                })?;
+                let signer1 = signers
+                    .get(signer_idx1)
```
Comment (majecty) [inline core/src/consensus/stake/actions.rs]: The signer may be different from the message's signer. We should get the prev term's signers.

File: core/src/consensus/stake/mod.rs [hunk 2]
```diff
@@ -428,6 +454,20 @@ pub fn ban(state: &mut TopLevelState, criminal: Address) -> StateResult<Deposit>
         (_, Some(jailed)) => jailed.deposit,
         _ => 0,
     };
+    // confiscate criminal's deposit and give the same deposit amount to the informant.
+    if let Some(criminal_info) = candidates.get_candidate(&criminal).map(Clone::clone) {
+        candidates.remove(&criminal);
+        if let Some(informant_info) = candidates.get_candidate(&public_to_address(informant)).map(Clone::clone) {
+            candidates.add_deposit(
```
Comment (majecty) [inline core/src/consensus/stake/mod.rs]: Why give CCC as a deposit?
Any reason for this? How about give CCC directly to the informant?

Comment (HoOngEe) [reply core/src/consensus/stake/mod.rs]: Yes you are right, I misread the spec

**Comment by HoOngEe** [comment]
@foriequal0 I remember you mentioned about the expected compile error 
```
^^^^^^ expected trait cstate::ActionHandler, found struct `consensus::stake::Stake`
```
at the [line](https://github.com/CodeChain-io/codechain/compare/2f4e11e713766d6ed3db25fecf9349bf04015be0..73fae8f8448539663fe04d74a28f8ee505c8f12f#diff-719e81628e55d2525e7149afab3dc1c0R99)
But I could not recall the solution.

**Comment by HoOngEe** [comment]
I have almost done. But I would better to change the error handling policy on "Cannot get validators from the parent state" part.

File: core/src/consensus/stake/actions.rs (hunk 1 above)
Comment (HoOngEe) [reply core/src/consensus/stake/actions.rs]: Fixed

**Comment by HoOngEe** [comment]
Now I'll change the error types which would be returned by the `ActionHandler::verify` method.

File: core/src/consensus/stake/mod.rs [hunk 3]
```diff
@@ -428,6 +458,11 @@ pub fn ban(state: &mut TopLevelState, criminal: Address) -> StateResult<Deposit>
         (_, Some(jailed)) => jailed.deposit,
         _ => 0,
     };
+    // confiscate criminal's deposit and give the same deposit amount to the informant.
+    candidates.remove(&criminal);
```
[the remark this answers is not in the document]
Comment (HoOngEe) [reply core/src/consensus/stake/mod.rs]: Thank you, I didn't catch it while rebasing

File: core/src/consensus/mod.rs [hunk 4]
```diff
@@ -273,6 +274,10 @@ pub trait ConsensusEngine: Sync + Send {
         self.action_handlers().iter().find(|handler| handler.handler_id() == id).map(AsRef::as_ref)
     }
 
+    fn get_validator_set(&self) -> Option<Arc<ValidatorSet>> {
```
[the remark this answers is not in the document]
Comment (HoOngEe) [reply core/src/consensus/mod.rs]: I forgot to remove it after it becomes useless.

File: core/src/consensus/validator_set/mod.rs [hunk 5]
```diff
@@ -45,6 +45,9 @@ pub trait ValidatorSet: Send + Sync {
     /// Draws a validator from nonce modulo number of validators.
     fn get_index(&self, parent: &H256, public: &Public) -> Option<usize>;
 
+    /// Draws whole validators' public key.
+    fn get_whole_publics(&self, parent: &H256) -> Vec<Public>;
```
[the remark this answers is not in the document]
Comment (HoOngEe) [reply core/src/consensus/validator_set/mod.rs]: Thanks

File: core/src/consensus/stake/actions.rs [hunk 6]
```diff
@@ -102,12 +116,59 @@ impl Action {
...
+                    })?;
+
+                let signer1 = signers_vec.get(signer_idx1).ok_or_else(|| {
+                    SyntaxError::InvalidCustomAction(String::from("Invalid signer index in message1"))
+                })?;
+                let signer2 = signers_vec.get(signer_idx2).ok_or_else(|| {
+                    SyntaxError::InvalidCustomAction(String::from("Invalid signer index in message2"))
+                })?;
+
+                verify_schnorr(signer1, &signature1, &message_hash)
```
[the remark this answers is not in the document]
Comment (HoOngEe) [reply core/src/consensus/stake/actions.rs]: Thank you very much

**Comment by majecty** [comment]
@foriequal0 Could you check this PR again?

File: core/src/consensus/solo/mod.rs [hunk 7]
```diff
@@ -37,14 +39,57 @@ pub struct Solo {
...
+
+    fn signer_index(&self) -> usize {
+        Default::default()
+    }
+
+    fn block_hash(&self) -> Option<H256> {
+        None
+    }
+
+    fn round(&self) -> &bool {
```
Comment (foriequal0) [inline core/src/consensus/solo/mod.rs]: 🤔 returning reference

File: core/src/consensus/stake/actions.rs [hunk 8]
```diff
@@ -102,12 +116,68 @@ impl Action {
...
+                    )))
+                }
+
+                let signer_idx1 = message1.signer_index();
+                let signer_idx2 = message2.signer_index();
+
+                let signed_block_hash = message1.block_hash();
+                let client = client.expect("Client should be initialized");
+                let parent_block_hash = signed_block_hash
+                    .and_then(|hash| client.block_header(&hash.into()))
```
Comment (foriequal0) [inline core/src/consensus/stake/actions.rs]: Can we sure that we can get a block header from `message1.block_hash()`? We don't know whether `message1` is a vote to a valid block or an invalid. When `message1` is a vote to an invalid block, We couldn't get the header. We should impose a limitation that `message1` should be a valid vote and `message2` is an invalid vote, or make them symmetric, or get header from height.

Comment (foriequal0) [reply core/src/consensus/stake/actions.rs]: I think this also missing the case
```
message1 != message2 
&& message1.round() == message2.round()
&& message1.block_hash() == message2.block_hash()
&& signer(message1) != signer(message2)
```
Same block, Same round, but different signer (obviously they are different message)

File: core/src/consensus/stake/mod.rs [hunk 9]
```diff
@@ -245,6 +262,20 @@ fn self_nominate(
...
+    state: &mut TopLevelState,
+    sender_public: &Public,
+    message: M,
+) -> StateResult<()> {
+    let idx = message.signer_index();
+    let validators = get_validators(state)?;
+    let validator_pubkeys: Vec<Public> = validators.into_iter().map(|val| *val.pubkey()).collect();
+    let pubkey = validator_pubkeys.get(idx).expect("Index was verified");
+
+    ban(state, sender_public, public_to_address(pubkey))?;
```
Comment (foriequal0) [inline core/src/consensus/stake/mod.rs]: You can avoid `collect` by
```rust
    let validator = get_validators(state)?.get(idx).expect("Index was verified");
    ban(state, sender_public, public_to_address(validator.pubkey()))?;
```

File: core/src/consensus/stake/actions.rs (hunk 8 above)
Comment (HoOngEe) [reply core/src/consensus/stake/actions.rs]: Thank you. I changed the code to get parent_block_hash based on the height information.

File: core/src/consensus/stake/mod.rs (hunk 9 above)
Comment (HoOngEe) [reply core/src/consensus/stake/mod.rs]: Thank you, I removed `collect`.

**Comment by HoOngEe** [comment]
Now I'm going to add a commit for unit testing

File: core/src/consensus/solo/mod.rs [hunk 10]
```diff
@@ -18,16 +18,18 @@ mod params;
 
 use std::sync::Arc;
 
-use ckey::Address;
+use ckey::{Address, Error as keyError, Public, SchnorrSignature};
```
Comment (foriequal0) [inline core/src/consensus/solo/mod.rs]:
```suggestion
use ckey::{Address, Error as KeyError, Public, SchnorrSignature};
```

**Comment by majecty** [comment]
@HoOngEe Please check the failed travis job.

File: core/src/consensus/stake/actions.rs [hunk 11]
```diff
@@ -247,16 +347,283 @@ mod tests {
...
+                signer_index: 0,
+            },
+            &|v| v,
+            &block_hash_twister,
+        );
+        let expected_err = Err(VerifyError::Syntax(SyntaxError::InvalidCustomAction(String::from(
+            "Schnorr signature verification fails",
+        ))));
+        assert_eq!(result, expected_err);
+    }
```
Comment (foriequal0) [inline core/src/consensus/stake/actions.rs]: Would you like to add a test that a message is from an older block?

File: core/src/consensus/tendermint/message.rs [hunk 12]
```diff
@@ -344,6 +347,7 @@ pub fn message_hash(step: VoteStep, block_hash: H256) -> H256 {
 
 #[cfg(test)]
 mod tests {
+    use ckey::SchnorrSignature;
```
[the remark this answers is not in the document]
Comment (HoOngEe) [reply core/src/consensus/tendermint/message.rs]: Thank you, I deleted it.

File: core/src/consensus/stake/actions.rs [hunk 13]
```diff
@@ -102,12 +116,72 @@ impl Action {
...
+                        "The messages are from two different voting rounds",
+                    ))
+                    .into())
+                }
+
+                let signer_idx1 = message1.signer_index();
+                let signer_idx2 = message2.signer_index();
+
+                // we have checked message1.round() == message2.round() so we can certain that
+                // message1.height() == message2.height(). So symmetry beaking does not occur here.
```
[the remark this answers is not in the document]
Comment (HoOngEe) [reply core/src/consensus/stake/actions.rs]: Thanks, that's clearly where the assertion is needed.

File: core/src/consensus/stake/actions.rs [hunk 14]
```diff
@@ -102,12 +116,72 @@ impl Action {
...
+                    .ok_or_else(|| {
+                        SyntaxError::InvalidCustomAction(format!(
+                            "Cannot get header from the height {}",
+                            signed_block_height
+                        ))
+                    })?
+                    .parent_hash();
+
+                let parent_state = client
+                    .state_at(parent_block_hash.into())
```
[the remark this answers is not in the document]
Comment (HoOngEe) [reply core/src/consensus/stake/actions.rs]: Ah, I did not recalled it. Ok, we can calculate blockId from the height.

File: core/src/consensus/stake/actions.rs [hunk 15]
```diff
@@ -102,12 +116,72 @@ impl Action {
...
+                let signer1 = signers_vec.get(signer_idx1).ok_or_else(|| RuntimeError::InvalidValidatorIndex {
+                    idx: signer_idx1,
+                    parent_block_hash,
+                })?;
+                let signer2 = signers_vec.get(signer_idx2).ok_or_else(|| RuntimeError::InvalidValidatorIndex {
+                    idx: signer_idx2,
+                    parent_block_hash,
+                })?;
+
+                if signer1 != signer2 {
```
[the remark this answers is not in the document]
Comment (HoOngEe) [reply core/src/consensus/stake/actions.rs]: So sharp. I'll change the code like this:
First compare `signer_idx1` and `signer_idx2`. If they are different, invoke an error.
Then, I'll get signer public key without collecting all pubkeys in vector.

File: core/src/consensus/stake/actions.rs [hunk 16]
```diff
@@ -102,12 +116,72 @@ impl Action {
...
+                            "Cannot get header from the height {}",
+                            signed_block_height
+                        ))
+                    })?
+                    .parent_hash();
+
+                let parent_state = client
+                    .state_at(parent_block_hash.into())
+                    .ok_or_else(|| TrieError::InvalidStateRoot(parent_block_hash))?;
+                let signers_vec: Vec<Public> = get_validators(&parent_state)
```
[the remark this answers is not in the document]
Comment (HoOngEe) [reply core/src/consensus/stake/actions.rs]: I changed this part to use validatorSet

File: core/src/consensus/stake/mod.rs [hunk 17]
```diff
@@ -245,6 +262,17 @@ fn self_nominate(
     Ok(())
 }
 
+pub fn ban_double_vote_account<M: Message>(
+    state: &mut TopLevelState,
+    sender_public: &Public,
+    message: M,
+) -> StateResult<()> {
+    let idx = message.signer_index();
+    let validator_pubkey = *get_validators(state)?.get(idx).expect("Index was verified").pubkey();
```
[the remark this answers is not in the document]
Comment (HoOngEe) [reply core/src/consensus/stake/mod.rs]: Yes It's clearly a fault.

File: core/src/consensus/stake/actions.rs (hunk 11 above)
Comment (HoOngEe) [reply core/src/consensus/stake/actions.rs]: I didn't catch the meaning of "a message from an older block" .
In this test I manually set the block height of the messages. So any block can be tested.

**Comment by HoOngEe** [comment]
I now need to reject reportDoubleVote for already banned account.

File: core/src/consensus/stake/mod.rs [hunk 18]
```diff
@@ -414,12 +443,13 @@ pub fn jail(state: &mut TopLevelState, addresses: &[Address], custody_until: u64
 }
 
 #[allow(dead_code)]
```
Comment (majecty) [inline core/src/consensus/stake/mod.rs]: Let's remove `#[allow(dead_code)]`

Comment (HoOngEe) [reply core/src/consensus/stake/mod.rs]: I removed it.

File: core/src/consensus/stake/actions.rs [hunk 19]
```diff
@@ -50,10 +56,19 @@ pub enum Action {
...
+    },
 }
 
-impl Action {
-    pub fn verify(&self, current_params: &CommonParams) -> Result<(), SyntaxError> {
+impl<M: Message> Action<M> {
+    pub fn verify(
+        &self,
+        current_params: &CommonParams,
+        client: Option<Arc<ConsensusClient>>,
```
Comment (foriequal0) [inline core/src/consensus/stake/actions.rs]: You could've accept just `&ConsensusClient`. Anyway, We'll ship it now and fix it later.

File: core/src/consensus/stake/actions.rs (hunk 11 above)
Comment (foriequal0) [reply core/src/consensus/stake/actions.rs]: I was referring that you should catch this kind of [bug](
https://github.com/CodeChain-io/codechain/pull/1640#discussion_r306205318) too. I didn't know that you are just test 'verify' function. Let it be handled in e2e test then.

**Pull request merged**